cpu/esp8266/Makefile.include

```diff
@@ -0,0 +1,175 @@
+# check some environment variables first
+ifndef NEWLIB_DIR
+    $(info NEWLIB_DIR should be defined as /path/to/newlib directory)
+    $(info NEWLIB_DIR is set by default to /opt/esp/newlib-xtensa)
+    export NEWLIB_DIR=/opt/esp/newlib-xtensa
+endif
+
+ifndef SDK_DIR
+    $(info SDK_DIR should be defined as /path/to/sdk directory)
+    $(info SDK_DIR is set by default to /opt/esp/esp-open-sdk/sdk)
+    export SDK_DIR=/opt/esp/esp-open-sdk/sdk
+endif
+
+export CPU ?= esp8266
+export TARGET_ARCH ?= xtensa-lx106-elf
+
+INCLUDES += -I$(NEWLIB_DIR)/$(TARGET_ARCH)/include
+INCLUDES += -I$(RIOTBOARD)/common/$(CPU)/include
+INCLUDES += -I$(RIOTCPU)/$(CPU)
+INCLUDES += -I$(RIOTCPU)/$(CPU)/espressif
+
+CFLAGS  += -DRIOT_OS -DESP_OPEN_SDK
+CFLAGS  += -Wno-unused-parameter -Wformat=0
+CFLAGS  += -mlongcalls -mtext-section-literals
+ASFLAGS += --longcalls --text-section-literals
+
+NETIF_NUMOF=0
+
+ifeq ($(SDK), 1)
+    INCLUDES += -I$(SDK_DIR)/include
+    CFLAGS   += -DSDK_USED -DUSE_US_TIMER
+    ifeq ($(SDK_INT_HANDLING), 1)
+        CFLAGS += -DSDK_INT_HANDLING
+    endif
+else
+    INCLUDES += -I$(SDK_DIR)/include
+    CFLAGS   += -DSDK_NOT_USED
+endif
+
+CFLAGS_OPT = -fzero-initialized-in-bss -O2
+
+ifeq ($(ENABLE_GDBSTUB), 1)
+    GDBSTUB_DIR ?= $(RIOTCPU)/$(CPU)/esp-gdbstub
+    CFLAGS_OPT   = -Og -ggdb -DGDBSTUB_FREERTOS=0
+    CFLAGS      += -DENABLE_GDBSTUB
+    INCLUDES    += -I$(GDBSTUB_DIR)
+    USEMODULE   += gdbstub
+endif
+
+ifeq ($(ENABLE_GDB), 1)
+    CFLAGS_OPT  = -fzero-initialized-in-bss -Og -ggdb
+endif
+
+CFLAGS += $(CFLAGS_OPT)
+
+ifeq ($(ENABLE_SW_TIMER), 1)
```

Comment:
IMO these ```ENABLE_*``` should be modeled with RIOT's module and feature system, and moved to the CPU's Makefile.dep, where appropriate.

Reply:
Sorry, I'm still in the process to understand the makefile structure, which is pretty complex. It is not always clear which on-board means are available to me.

I simply introduced the ```ENABLE_*``` make variables to have a fast possibility to switch on/off very special compile options without changing the makefile everytime.

For example, ```ENABLE_SW_TIMER``` decides whether to compile the ```periph/timer``` module with hardware timers or with software timers. By default, hardware timers are used. The ```FEATURES_PROVIDED += periph_timer``` does not provide the possibility to divide between hardware or software timers.

Another example is ```ENABLE_MRF24J40```. I was wondering how to tell ```examples/gnrc_networking``` to use module ```mrf24j40``` and to activate the SPI interface. I could not figure out a way, how to activate SPI. Even though each board has the SPI interface, it should be activated only if it is really required. ESP8266 has only 9 available GPIOs. With SPI activated only 5 GPIOs are left :-(

Reply:
@kaspar30 Ah ok, I see, ```MODULE_PERIPH_SPI``` could be used instead of ```SPI_USED``` to activate SPI. 
But the question how to tell module ```periph/timer``` to use software timers instead of hardware timers and how to tell ```examples/gnrc_networking``` to use module ```mrf24j40``` are still open.

Reply:
> I'm still in the process to understand the makefile structure

That's totally alright!

> MODULE_PERIPH_SPI could be used instead of SPI_USED to activate SPI.

Exactly.

> But the question how to tell module periph/timer to use software timers instead of hardware timers 

We sometimes use "pseudomodules" for these boolean flag type configuration options, as they can then be depended on by any module. If that is not needed, I think we can keep the define for now.

> how to tell examples/gnrc_networking to use module mrf24j40 are still open.

Take a look at how samr21-xpro includes at86rf233 in ```boards/samr21-xpro/Makefile.dep```.
Basically, add mrf24j40 to USEMODULE if netdev_default is in USEMODULE.
That way, mrf24j40 will be pulled in whenever a network stack is selected.

Reply:
> Take a look at how samr21-xpro includes at86rf233 in boards/samr21-xpro/Makefile.dep.
> Basically, add mrf24j40 to USEMODULE if netdev_default is in USEMODULE.
> That way, mrf24j40 will be pulled in whenever a network stack is selected.

Hm, but what if I want to compile the same application one time with ```mrf24j40```, another time with ```enc28j60``` and third time with the WiFi built in netdev ```espnow```?

What would be the right way to enable debugging in RIOT and optionally add gdb stub code (```ENABLE_GDB``` and ```ENABLE_GDBSTUB```)?

Reply:
> Hm, but what if I want to compile the same application one time with mrf24j40, another time with enc28j60 and third time with the WiFi built in netdev espnow?

Ah, sorry, I thought mrf24j40 was an onboard chip? If not, why mention it at all in the board configuration?